File: examples/knowledge_base_demo.py
"""
Knowledge Base Demo
Demonstrates loading and using the CARETALE AI knowledge base
"""

import sys
from pathlib import Path

# Add src directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent / "python" / "src"))

from rag_system.knowledge_base_loader import create_and_load_knowledge_base
from rag_system.knowledge_base_manager import KnowledgeBaseCategory
from models.knowledge_document import MedicalDomain


def main():
    print("=== CARETALE AI Knowledge Base Demo ===\n")

    # Load the knowledge base (adjust path to be relative to workspace root)
    print("Loading knowledge base...")
    data_path = str(Path(__file__).parent.parent / "data" / "knowledge-base")
    knowledge_base = create_and_load_knowledge_base(data_path)
    print()

    # Display statistics
    print("Knowledge Base Statistics:")
    stats = knowledge_base.get_statistics()
    print(f"  Total documents: {stats['total_documents']}")
    print(f"  Validated: {stats['validated']}")
    print(f"  Pending review: {stats['pending_review']}")
    print(f"  Outdated: {stats['outdated']}")
    print()

    print("Documents by Category:")
    print(f"  Clinical Guidelines: {stats['category_clinical_guidelines']}")
    print(f"  Educational Content: {stats['category_educational_content']}")
    print(f"  Drug Information: {stats['category_drug_information']}")
    print(f"  Care Protocols: {stats['category_care_protocols']}")
    print()

    # Get clinical guidelines
    print("=== Clinical Guidelines ===")
    guidelines = knowledge_base.get_documents_by_category(
        KnowledgeBaseCategory.CLINICAL_GUIDELINES
    )
    for doc in guidelines:
        print(f"\n{doc.metadata.title}")
        print(f"  Source: {doc.source.name}")
        print(f"  Domain: {doc.domain.value}")
        print(f"  Last Updated: {doc.last_updated.date()}")
        print(f"  Citation: {doc.citation_format.short_form}")
    print()

    # Get drug information
    print("=== Drug Information ===")
    drugs = knowledge_base.get_documents_by_category(
        KnowledgeBaseCategory.DRUG_INFORMATION
    )
    for doc in drugs:
        print(f"\n{doc.metadata.title}")
        print(f"  Source: {doc.source.name}")
        print(f"  Keywords: {', '.join(doc.metadata.keywords)}")
    print()

    # Get documents by domain
    print("=== Cardiology Documents ===")
    cardiology_docs = knowledge_base.get_documents_by_domain(MedicalDomain.CARDIOLOGY)
    print(f"Found {len(cardiology_docs)} cardiology documents:")
    for doc in cardiology_docs:
        print(f"  - {doc.metadata.title} ({doc.source.authority})")
    print()

    # Get a specific document
    print("=== Specific Document Example ===")
    heart_failure_doc = knowledge_base.get_document("cg-001-heart-failure")
    if heart_failure_doc:
        print(f"Title: {heart_failure_doc.metadata.title}")
        print(f"Source: {heart_failure_doc.source.name}")
        print("\nContent Preview (first 200 chars):")
        print(heart_failure_doc.content[:200] + "...")
        print("\nFull Citation:")
        print(heart_failure_doc.citation_format.apa)
    print()

    # Check for outdated documents
    print("=== Document Maintenance ===")
    outdated = knowledge_base.get_outdated_documents(365)
    if outdated:
        print(f"Found {len(outdated)} documents older than 365 days:")
        for doc in outdated:
            from datetime import datetime, timezone
            age = (datetime.now(timezone.utc) - doc.last_updated).days
            print(f"  - {doc.metadata.title} ({age} days old)")
    else:
        print("All documents are up to date (less than 365 days old)")
    print()

    print("=== Demo Complete ===")


if __name__ == "__main__":
    main()

File: examples/knowledge-base-demo.ts
/**
 * Knowledge Base Demo
 * Demonstrates loading and using the CARETALE AI knowledge base
 */

import { createAndLoadKnowledgeBase } from '../src/knowledge-context-layer/knowledge-base-loader';
import { KnowledgeBaseCategory } from '../src/knowledge-context-layer/types';
import { MedicalDomain } from '../src/shared/types/knowledge-document';

function main() {
  console.log('=== CARETALE AI Knowledge Base Demo ===\n');

  // Load the knowledge base
  console.log('Loading knowledge base...');
  const knowledgeBase = createAndLoadKnowledgeBase();
  console.log('');

  // Display statistics
  console.log('Knowledge Base Statistics:');
  const stats = knowledgeBase.getStatistics();
  console.log(`  Total documents: ${stats.total_documents}`);
  console.log(`  Validated: ${stats.validated}`);
  console.log(`  Pending review: ${stats.pending_review}`);
  console.log(`  Outdated: ${stats.outdated}`);
  console.log('');

  console.log('Documents by Category:');
  console.log(`  Clinical Guidelines: ${stats.category_clinical_guidelines}`);
  console.log(`  Educational Content: ${stats.category_educational_content}`);
  console.log(`  Drug Information: ${stats.category_drug_information}`);
  console.log(`  Care Protocols: ${stats.category_care_protocols}`);
  console.log('');

  // Get clinical guidelines
  console.log('=== Clinical Guidelines ===');
  const guidelines = knowledgeBase.getDocumentsByCategory(KnowledgeBaseCategory.CLINICAL_GUIDELINES);
  guidelines.forEach((doc) => {
    console.log(`\n${doc.metadata.title}`);
    console.log(`  Source: ${doc.source.name}`);
    console.log(`  Domain: ${doc.domain}`);
    console.log(`  Last Updated: ${doc.lastUpdated.toISOString().split('T')[0]}`);
    console.log(`  Citation: ${doc.citationFormat.shortForm}`);
  });
  console.log('');

  // Get drug information
  console.log('=== Drug Information ===');
  const drugs = knowledgeBase.getDocumentsByCategory(KnowledgeBaseCategory.DRUG_INFORMATION);
  drugs.forEach((doc) => {
    console.log(`\n${doc.metadata.title}`);
    console.log(`  Source: ${doc.source.name}`);
    console.log(`  Keywords: ${doc.metadata.keywords.join(', ')}`);
  });
  console.log('');

  // Get documents by domain
  console.log('=== Cardiology Documents ===');
  const cardiologyDocs = knowledgeBase.getDocumentsByDomain(MedicalDomain.CARDIOLOGY);
  console.log(`Found ${cardiologyDocs.length} cardiology documents:`);
  cardiologyDocs.forEach((doc) => {
    console.log(`  - ${doc.metadata.title} (${doc.source.authority})`);
  });
  console.log('');

  // Get a specific document
  console.log('=== Specific Document Example ===');
  const heartFailureDoc = knowledgeBase.getDocument('cg-001-heart-failure');
  if (heartFailureDoc) {
    console.log(`Title: ${heartFailureDoc.metadata.title}`);
    console.log(`Source: ${heartFailureDoc.source.name}`);
    console.log(`\nContent Preview (first 200 chars):`);
    console.log(heartFailureDoc.content.substring(0, 200) + '...');
    console.log(`\nFull Citation:`);
    console.log(heartFailureDoc.citationFormat.apa);
  }
  console.log('');

  // Check for outdated documents
  console.log('=== Document Maintenance ===');
  const outdated = knowledgeBase.getOutdatedDocuments(365);
  if (outdated.length > 0) {
    console.log(`Found ${outdated.length} documents older than 365 days:`);
    outdated.forEach((doc) => {
      const age = Math.floor(
        (Date.now() - doc.lastUpdated.getTime()) / (1000 * 60 * 60 * 24)
      );
      console.log(`  - ${doc.metadata.title} (${age} days old)`);
    });
  } else {
    console.log('All documents are up to date (less than 365 days old)');
  }
  console.log('');

  console.log('=== Demo Complete ===');
}

// Run the demo
main();
